### ISSUE TITLE
Launch new containers will lead all bird route to be removed first

### ISSUE THREAD

**Issue by frostynova** [Metaswitch/calico-docker#136]
I found after adding new containers, route rules added by bird on other hosts' will  be removed first, then added back.

This will cause ```ping``` lose about 7 packets

Our host: 192.168.77.0/24
Calico pool: 172.20.0.0/16
Environment: Vagrant + CentOS 7.0 + Docker 1.5 + Calico master branch

bird.log

```
2015-05-19 04:13:02 <WARN> Netlink: File exists
2015-05-19 04:13:02 <WARN> Netlink: File exists
2015-05-19 04:13:02 <WARN> ...
2015-05-19 04:13:02 <INFO> Reconfiguration requested by SIGHUP
2015-05-19 04:13:02 <INFO> Reconfiguring
2015-05-19 04:13:02 <TRACE> direct1: Reconfigured
2015-05-19 04:13:02 <INFO> Removing protocol bgp1
2015-05-19 04:13:02 <TRACE> bgp1: Shutting down
2015-05-19 04:13:02 <TRACE> bgp1: Shutdown requested
2015-05-19 04:13:02 <TRACE> bgp1: State changed to stop
2015-05-19 04:13:02 <TRACE> bgp1: BGP session closed
2015-05-19 04:13:02 <INFO> Adding protocol bgp1
2015-05-19 04:13:02 <TRACE> bgp1: Initializing
2015-05-19 04:13:02 <TRACE> bgp1: Starting
2015-05-19 04:13:02 <TRACE> bgp1: State changed to start
2015-05-19 04:13:02 <TRACE> bgp1: Sending NOTIFICATION(code=6.3)
2015-05-19 04:13:02 <TRACE> bgp1: Down
2015-05-19 04:13:02 <TRACE> bgp1: State changed to down
2015-05-19 04:13:02 <TRACE> bgp1: Started
2015-05-19 04:13:02 <TRACE> bgp1: Connect delayed by 5 seconds
2015-05-19 04:13:02 <INFO> Reconfigured
2015-05-19 04:13:02 <INFO> Reconfiguration requested by SIGHUP
2015-05-19 04:13:02 <INFO> Reconfiguring
2015-05-19 04:13:02 <TRACE> direct1: Reconfigured
2015-05-19 04:13:02 <INFO> Removing protocol bgp1
2015-05-19 04:13:02 <TRACE> bgp1: Shutting down
2015-05-19 04:13:02 <TRACE> bgp1: Shutdown requested
2015-05-19 04:13:02 <TRACE> bgp1: State changed to stop
2015-05-19 04:13:02 <INFO> Adding protocol bgp1
2015-05-19 04:13:02 <TRACE> bgp1: Initializing
2015-05-19 04:13:02 <TRACE> bgp1: Starting
2015-05-19 04:13:02 <TRACE> bgp1: State changed to start
2015-05-19 04:13:02 <TRACE> bgp1: Down
2015-05-19 04:13:02 <TRACE> bgp1: State changed to down
2015-05-19 04:13:02 <TRACE> bgp1: Started
2015-05-19 04:13:02 <TRACE> bgp1: Connect delayed by 5 seconds
2015-05-19 04:13:02 <INFO> Reconfigured
2015-05-19 04:13:03 <INFO> Reconfiguration requested by SIGHUP
2015-05-19 04:13:03 <INFO> Reconfiguring
2015-05-19 04:13:03 <TRACE> direct1: Reconfigured
2015-05-19 04:13:03 <INFO> Removing protocol bgp1
2015-05-19 04:13:03 <TRACE> bgp1: Shutting down
2015-05-19 04:13:03 <TRACE> bgp1: Shutdown requested
2015-05-19 04:13:03 <TRACE> bgp1: State changed to stop
2015-05-19 04:13:03 <INFO> Adding protocol bgp1
2015-05-19 04:13:03 <TRACE> bgp1: Initializing
2015-05-19 04:13:03 <TRACE> bgp1: Starting
2015-05-19 04:13:03 <TRACE> bgp1: State changed to start
2015-05-19 04:13:03 <TRACE> bgp1: Down
2015-05-19 04:13:03 <TRACE> bgp1: State changed to down
2015-05-19 04:13:03 <TRACE> bgp1: Started
2015-05-19 04:13:03 <TRACE> bgp1: Connect delayed by 5 seconds
2015-05-19 04:13:03 <INFO> Reconfigured
2015-05-19 04:13:04 <WARN> Netlink: File exists
2015-05-19 04:13:04 <WARN> Netlink: File exists
2015-05-19 04:13:04 <WARN> ...
2015-05-19 04:13:06 <WARN> Netlink: File exists
2015-05-19 04:13:06 <WARN> Netlink: File exists
2015-05-19 04:13:06 <WARN> ...
2015-05-19 04:13:07 <TRACE> bgp1: Connecting to 192.168.77.21 from local address 192.168.77.22
2015-05-19 04:13:07 <TRACE> bgp1: Connected
2015-05-19 04:13:07 <TRACE> bgp1: Sending OPEN(ver=4,as=64511,hold=240,id=c0a84d16)
2015-05-19 04:13:07 <TRACE> bgp1: Got OPEN(as=64511,hold=240,id=c0a84d15)
2015-05-19 04:13:07 <TRACE> bgp1: Sending KEEPALIVE
2015-05-19 04:13:07 <TRACE> bgp1: Got KEEPALIVE
2015-05-19 04:13:07 <TRACE> bgp1: BGP session established
2015-05-19 04:13:07 <TRACE> bgp1: Connected to table master
2015-05-19 04:13:07 <TRACE> bgp1: State changed to feed
2015-05-19 04:13:07 <TRACE> bgp1 < filtered out 0.0.0.0/0 via 10.0.2.2 on enp0s3
2015-05-19 04:13:07 <TRACE> bgp1 < filtered out 10.0.2.0/24 dev enp0s3
2015-05-19 04:13:07 <TRACE> bgp1 < filtered out 192.168.77.0/24 dev enp0s8
2015-05-19 04:13:07 <TRACE> bgp1 < filtered out 169.254.0.0/16 dev enp0s8
2015-05-19 04:13:07 <TRACE> bgp1 < filtered out 172.17.0.0/16 dev docker0
2015-05-19 04:13:07 <TRACE> bgp1: State changed to up
2015-05-19 04:13:07 <TRACE> bgp1: Sending END-OF-RIB
2015-05-19 04:13:07 <TRACE> bgp1: Got UPDATE
2015-05-19 04:13:07 <TRACE> bgp1 > added [best] 172.20.0.1/32 via 192.168.77.21 on enp0s8
2015-05-19 04:13:07 <TRACE> bgp1 < rejected by protocol 172.20.0.1/32 via 192.168.77.21 on enp0s8
2015-05-19 04:13:07 <TRACE> bgp1: Got UPDATE
2015-05-19 04:13:07 <TRACE> bgp1: Got END-OF-RIB
2015-05-19 04:13:08 <WARN> Netlink: File exists
2015-05-19 04:13:08 <WARN> Netlink: File exists
2015-05-19 04:13:08 <WARN> ...
2015-05-19 04:13:10 <WARN> Netlink: File exists
2015-05-19 04:13:10 <WARN> Netlink: File exists
2015-05-19 04:13:10 <WARN> ...
2015-05-19 04:13:12 <WARN> Netlink: File exists
2015-05-19 04:13:12 <WARN> Netlink: File exists
2015-05-19 04:13:12 <WARN> ...
2015-05-19 04:13:14 <WARN> Netlink: File exists
2015-05-19 04:13:14 <WARN> Netlink: File exists
2015-05-19 04:13:14 <WARN> ...
2015-05-19 04:13:16 <WARN> Netlink: File exists
2015-05-19 04:13:16 <WARN> Netlink: File exists
2015-05-19 04:13:16 <WARN> ...
2015-05-19 04:13:18 <WARN> Netlink: File exists
2015-05-19 04:13:18 <WARN> Netlink: File exists
2015-05-19 04:13:18 <WARN> ...
2015-05-19 04:13:20 <WARN> Netlink: File exists
2015-05-19 04:13:20 <WARN> Netlink: File exists
2015-05-19 04:13:20 <WARN> ...
2015-05-19 04:13:21 <INFO> Reconfiguration requested by SIGHUP
2015-05-19 04:13:21 <INFO> Reconfiguring
2015-05-19 04:13:21 <TRACE> direct1: Reconfigured
2015-05-19 04:13:21 <INFO> Removing protocol bgp1
2015-05-19 04:13:21 <TRACE> bgp1: Shutting down
2015-05-19 04:13:21 <TRACE> bgp1: Shutdown requested
2015-05-19 04:13:21 <TRACE> bgp1: State changed to stop
2015-05-19 04:13:21 <TRACE> bgp1: BGP session closed
2015-05-19 04:13:21 <INFO> Adding protocol bgp1
2015-05-19 04:13:21 <TRACE> bgp1: Initializing
2015-05-19 04:13:21 <TRACE> bgp1: Starting
2015-05-19 04:13:21 <TRACE> bgp1: State changed to start
2015-05-19 04:13:21 <TRACE> bgp1 > removed [sole] 172.20.0.1/32 via 192.168.77.21 on enp0s8
2015-05-19 04:13:21 <TRACE> bgp1: Sending NOTIFICATION(code=6.3)
2015-05-19 04:13:21 <TRACE> bgp1: Down
2015-05-19 04:13:21 <TRACE> bgp1: State changed to down
2015-05-19 04:13:21 <TRACE> bgp1: Started
2015-05-19 04:13:21 <TRACE> bgp1: Connect delayed by 5 seconds
2015-05-19 04:13:21 <INFO> Reconfigured
2015-05-19 04:13:21 <INFO> Reconfiguration requested by SIGHUP
2015-05-19 04:13:21 <INFO> Reconfiguring
2015-05-19 04:13:21 <TRACE> direct1: Reconfigured
2015-05-19 04:13:21 <INFO> Removing protocol bgp1
2015-05-19 04:13:21 <TRACE> bgp1: Shutting down
2015-05-19 04:13:21 <TRACE> bgp1: Shutdown requested
2015-05-19 04:13:21 <TRACE> bgp1: State changed to stop
2015-05-19 04:13:21 <INFO> Adding protocol bgp1
2015-05-19 04:13:21 <TRACE> bgp1: Initializing
2015-05-19 04:13:21 <TRACE> bgp1: Starting
2015-05-19 04:13:21 <TRACE> bgp1: State changed to start
2015-05-19 04:13:21 <TRACE> bgp1: Down
2015-05-19 04:13:21 <TRACE> bgp1: State changed to down
2015-05-19 04:13:21 <TRACE> bgp1: Started
2015-05-19 04:13:21 <TRACE> bgp1: Connect delayed by 5 seconds
2015-05-19 04:13:21 <INFO> Reconfigured
2015-05-19 04:13:22 <WARN> Netlink: File exists
2015-05-19 04:13:22 <WARN> Netlink: File exists
2015-05-19 04:13:22 <WARN> ...
2015-05-19 04:13:24 <WARN> Netlink: File exists
2015-05-19 04:13:24 <WARN> Netlink: File exists
2015-05-19 04:13:24 <WARN> ...
2015-05-19 04:13:24 <TRACE> bgp1: Incoming connection from 192.168.77.21 (port 60303) accepted
2015-05-19 04:13:24 <TRACE> bgp1: Sending OPEN(ver=4,as=64511,hold=240,id=c0a84d16)
2015-05-19 04:13:26 <TRACE> bgp1: Connecting to 192.168.77.21 from local address 192.168.77.22
2015-05-19 04:13:26 <WARN> Netlink: File exists
2015-05-19 04:13:26 <WARN> Netlink: File exists
2015-05-19 04:13:26 <WARN> ...
2015-05-19 04:13:26 <TRACE> bgp1: Connected
2015-05-19 04:13:26 <TRACE> bgp1: Got OPEN(as=64511,hold=240,id=c0a84d15)
2015-05-19 04:13:26 <TRACE> bgp1: Sending OPEN(ver=4,as=64511,hold=240,id=c0a84d16)
2015-05-19 04:13:26 <TRACE> bgp1: Sending KEEPALIVE
2015-05-19 04:13:26 <TRACE> bgp1: Got KEEPALIVE
2015-05-19 04:13:26 <TRACE> bgp1: BGP session established
2015-05-19 04:13:26 <TRACE> bgp1: Connected to table master
2015-05-19 04:13:26 <TRACE> bgp1: State changed to feed
2015-05-19 04:13:26 <RMT> bgp1: Received: Connection collision resolution
2015-05-19 04:13:26 <TRACE> bgp1: BGP session closed
2015-05-19 04:13:26 <TRACE> bgp1: State changed to stop
2015-05-19 04:13:26 <TRACE> bgp1: Sending NOTIFICATION(code=6.0)
2015-05-19 04:13:26 <TRACE> bgp1: Down
2015-05-19 04:13:26 <TRACE> bgp1: State changed to down
2015-05-19 04:13:26 <TRACE> bgp1: Starting
2015-05-19 04:13:26 <TRACE> bgp1: State changed to start
2015-05-19 04:13:26 <TRACE> bgp1: Started
2015-05-19 04:13:26 <TRACE> bgp1: Connect delayed by 5 seconds
2015-05-19 04:13:28 <WARN> Netlink: File exists
2015-05-19 04:13:28 <WARN> Netlink: File exists
2015-05-19 04:13:28 <WARN> ...
2015-05-19 04:13:30 <WARN> Netlink: File exists
2015-05-19 04:13:30 <WARN> Netlink: File exists
2015-05-19 04:13:30 <WARN> ...
2015-05-19 04:13:31 <TRACE> bgp1: Connecting to 192.168.77.21 from local address 192.168.77.22
2015-05-19 04:13:31 <TRACE> bgp1: Connected
2015-05-19 04:13:31 <TRACE> bgp1: Sending OPEN(ver=4,as=64511,hold=240,id=c0a84d16)
2015-05-19 04:13:31 <TRACE> bgp1: Got OPEN(as=64511,hold=240,id=c0a84d15)
2015-05-19 04:13:31 <TRACE> bgp1: Sending KEEPALIVE
2015-05-19 04:13:31 <TRACE> bgp1: Incoming connection from 192.168.77.21 (port 50108) accepted
2015-05-19 04:13:31 <TRACE> bgp1: Sending OPEN(ver=4,as=64511,hold=240,id=c0a84d16)
2015-05-19 04:13:31 <TRACE> bgp1: Got KEEPALIVE
2015-05-19 04:13:31 <TRACE> bgp1: BGP session established
2015-05-19 04:13:31 <TRACE> bgp1: Connected to table master
2015-05-19 04:13:31 <TRACE> bgp1: State changed to feed
2015-05-19 04:13:31 <TRACE> bgp1: Got OPEN(as=64511,hold=240,id=c0a84d15)
2015-05-19 04:13:31 <TRACE> bgp1: Connection collision, giving up this connection
2015-05-19 04:13:31 <TRACE> bgp1 < filtered out 0.0.0.0/0 via 10.0.2.2 on enp0s3
2015-05-19 04:13:31 <TRACE> bgp1 < filtered out 10.0.2.0/24 dev enp0s3
2015-05-19 04:13:31 <TRACE> bgp1 < filtered out 192.168.77.0/24 dev enp0s8
2015-05-19 04:13:31 <TRACE> bgp1 < filtered out 169.254.0.0/16 dev enp0s8
2015-05-19 04:13:31 <TRACE> bgp1 < filtered out 172.17.0.0/16 dev docker0
2015-05-19 04:13:31 <TRACE> bgp1: State changed to up
2015-05-19 04:13:31 <TRACE> bgp1: Sending NOTIFICATION(code=6.7)
2015-05-19 04:13:31 <TRACE> bgp1: Sending END-OF-RIB
2015-05-19 04:13:31 <TRACE> bgp1: Got UPDATE
2015-05-19 04:13:31 <TRACE> bgp1 > added [best] 172.20.0.2/32 via 192.168.77.21 on enp0s8
2015-05-19 04:13:31 <TRACE> bgp1 < rejected by protocol 172.20.0.2/32 via 192.168.77.21 on enp0s8
2015-05-19 04:13:31 <TRACE> bgp1 > added [best] 172.20.0.1/32 via 192.168.77.21 on enp0s8
2015-05-19 04:13:31 <TRACE> bgp1 < rejected by protocol 172.20.0.1/32 via 192.168.77.21 on enp0s8
2015-05-19 04:13:31 <TRACE> bgp1: Got UPDATE
2015-05-19 04:13:31 <TRACE> bgp1: Got END-OF-RIB
2015-05-19 04:13:32 <WARN> Netlink: File exists
2015-05-19 04:13:32 <WARN> Netlink: File exists
2015-05-19 04:13:32 <WARN> ...
2015-05-19 04:13:34 <WARN> Netlink: File exists
```

**Comment by frostynova** [comment]
After some trying, i found that ```pkill -HUP bird``` which in ```conf.d/bird.toml``` lead to this

**Comment by Symmetric** [comment]
From the Bird RPM [docs](http://rpm.pbone.net/index.php3/stat/4/idpl/23373923/dir/centos_6/com/bird-1.3.11-1.2.x86_64.rpm.html):

> * Soft reconfiguration -- no online commands for changing the configuration in very limited ways, just edit the configuration file and issue a `configure\' command or send SIGHUP and BIRD will start using the new configuration, possibly restarting protocols affected by the configuration changes

The docs don't mention protocol restarts though, and we'd assumed that this was safe. Should follow up with the Bird folks on this, might be a bug.

**Comment by frostynova** [comment]
@Symmetric Maybe ```graceful restart``` option of BGP protocol will help?

http://bird.network.cz/?get_doc&f=bird-6.html#ss6.2

**Comment by robbrockbank** [comment]
@frostynova -  I was digging in to this yesterday and think there may be a few things at fault.  One of which was the graceful restart option which I've added to my local branch and will be adding to master later today.  Also, following your suggestion of adding a prefix to minimize confd invocations and finally removing the timestamp from the bird templates since I suspect (although haven't confirmed just yet) that this is causing a BIRD config reload even though the template has not actually changed ... but the timestamp has so it makes it look like the bird configs are different.

Having graceful restart enabled in BIRD is going to help, especially in the case of a single BIRD instance failing.  It is less clear whether it will help in the full mesh scenario (rather than using a Route Reflector) because any addition or removal of a node will result in potentially simultaneous updates of BIRD across all nodes -- and graceful restart may not work in that scenario.  I'm looking into this a little more today.

**Comment by robbrockbank** [comment]
Adding a new container should no longer cause this.  Fix is in the latest master and will be in the next release.

Main fix here is to ensure we don't restart BIRD if the template hasn't changed.  Main issue was that the template contained a timestamp, so regeneration of the template made it appear to have changed.

Fixed by #141
